### PR TITLE
Implement layout constraint conversion to internal solver constraints

### PR DESCRIPTION
**Pull request by BennyKJohnson**
This PR implements the conversion of an NSLayoutConstraint into one or many internal solver constraints.

### CODE REVIEW

File: Headers/AppKit/NSView.h [hunk 1]
```diff
@@ -123,6 +124,9 @@ typedef enum _NSFocusRingType {
   NSFocusRingTypeExterior = 2
 } NSFocusRingType;
 
+extern const CGFloat NSViewNoInstrinsicMetric;
```
Comment (fredkiefer) [inline Headers/AppKit/NSView.h:127]: Why do we need this misspelled version? And would a `#define` be enough here?

File: Headers/AppKit/NSView.h [hunk 2]
```diff
@@ -672,10 +679,61 @@ PACKAGE_SCOPE
...
+@end
+
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_7, GS_API_LATEST)
+@interface NSView (NSConstraintBasedLayoutCoreMethods)
+
+- (void) updateConstraints;
+
+- (void) updateConstraintsForSubtreeIfNeeded;
+
+- (NSLayoutPriority)contentHuggingPriorityForOrientation:(NSLayoutConstraintOrientation)orientation;
```
Comment (fredkiefer) [inline Headers/AppKit/NSView.h:719]: You could use a few more spaces here to be consistent with your other methods. I know that the GNustep coding guidelines aren't that consistent by themselves. At least within a block we could try :-)

File: Source/NSView.m [hunk 3]
```diff
@@ -5242,6 +5273,145 @@ - (void) _layoutEngineDidChangeAlignmentRect
...
+    } else {
+      _huggingPriorities.vertical = priority;
+    }
+}
+
+@end
+
+@implementation NSView (NSConstraintBasedLayoutInstallingConstraints)
+
+- (GSAutoLayoutEngine*) _getOrCreateLayoutEngine
```
Comment (fredkiefer) [inline Source/NSView.m:5346]: Maybe move the method `_layoutEngine` to here as well?

**Review by fredkiefer — CHANGES_REQUESTED** [review]
I must admit, I have a bit of a problem following some of the methods on the GSAutoLayoutEngine. Perhaps a bit of rework there would help?

**Comment by BennyKJohnson** [comment]
> I must admit, I have a bit of a problem following some of the methods on the GSAutoLayoutEngine. Perhaps a bit of rework there would help?

Appreciate your feedback, and I agree that there are areas that could be improved upon in this class. Identifying the specific areas might help me find low hanging fruit that I can fix.

File: Source/GSAutoLayoutEngine.m [hunk 4]
```diff
@@ -199,8 +352,296 @@ - (void) removeConstraints: (NSArray *)constraints
...
+      return -1;
+    default:
+      return 1;
+    }
+}
+
+- (GSCSVariable *) variableForView: (NSView *)view
+                  andViewAttribute: (GSLayoutViewAttribute)attribute
+{
+  GSCSVariable *existingVariable =
```
[the remark this answers is not in the document]
Comment (BennyKJohnson) [reply Source/GSAutoLayoutEngine.m]: Great suggestion, fixed, this allowed me to simplify the code also.

File: Source/GSAutoLayoutEngine.m [hunk 5]
```diff
@@ -199,8 +352,296 @@ - (void) removeConstraints: (NSArray *)constraints
...
+- (GSCSVariable *) getExistingVariableForView: (NSView *)view
+                            withViewAttribute: (GSLayoutViewAttribute)attribute
+{
+  NSString *variableIdentifier =
+    [self getDynamicVariableIdentifierForView: view
+                            withViewAttribute: attribute];
+  return [self varibleWithName: variableIdentifier];
+}
+
+- (GSCSVariable *) varibleWithName: (NSString *)variableName
```
[the remark this answers is not in the document]
Comment (BennyKJohnson) [reply Source/GSAutoLayoutEngine.m]: Fixed

File: Source/NSView.m [hunk 6]
```diff
@@ -5242,6 +5273,145 @@ - (void) _layoutEngineDidChangeAlignmentRect
...
+- (void)setContentCompressionResistancePriority:(NSLayoutPriority)priority forOrientation:(NSLayoutConstraintOrientation)orientation {
+    if (orientation == NSLayoutConstraintOrientationHorizontal) {
+      _compressionPriorities.horizontal = priority;
+    } else {
+      _compressionPriorities.vertical = priority;
+    }
+}
+
+- (NSLayoutPriority)contentHuggingPriorityForOrientation:(NSLayoutConstraintOrientation)orientation {
+  GSIntrinsicContentSizePriority priorities = _huggingPriorities;
```
[the remark this answers is not in the document]
Comment (BennyKJohnson) [reply Source/NSView.m]: Fixed

File: Source/NSView.m [hunk 7]
```diff
@@ -5242,6 +5273,145 @@ - (void) _layoutEngineDidChangeAlignmentRect
...
+                                    frame: [self frame]
+                                superitem: [self superview]
+                                   bounds: [[self superview] bounds]];
+      [self addConstraints: autoresizingConstraints];
+    }
+
+  [self _setNeedsUpdateConstraints: NO];
+}
+
+- (NSLayoutPriority)contentCompressionResistancePriorityForOrientation:(NSLayoutConstraintOrientation)orientation {
```
[the remark this answers is not in the document]
Comment (BennyKJohnson) [reply Source/NSView.m]: Fixed

File: Source/NSView.m [hunk 8]
```diff
@@ -5202,8 +5206,35 @@ - (BOOL) translatesAutoresizingMaskIntoConstraints
...
+{
+  _contentCompressionResistancePriority = priority;
+}
+
+- (NSSize) intrinsicContentSize
+{
+  return NSMakeSize(NSViewNoIntrinsicMetric, NSViewNoIntrinsicMetric); 
+}
+
+-(CGFloat)baselineOffsetFromBottom
```
[the remark this answers is not in the document]
Comment (BennyKJohnson) [reply Source/NSView.m]: Fixed

File: Headers/AppKit/NSView.h (hunk 2 above)
Comment (BennyKJohnson) [reply Headers/AppKit/NSView.h:719]: Fixed

File: Headers/AppKit/NSView.h (hunk 1 above)
Comment (BennyKJohnson) [reply Headers/AppKit/NSView.h:127]: It was baked into the original header file and fixed in a new version. So we need to support both versions.

File: Source/GSAutoLayoutEngine.m [hunk 9]
```diff
@@ -79,6 +111,127 @@ - (NSRect) currentAlignmentRectForViewAtIndex: (NSNumber *)viewIndex;
...
+
+- (void) addInternalBaselineConstraintsForView: (NSView *)view
+                                    constraint: (GSCSConstraint *)constraint;
+
+- (void) addInternalSolverConstraint: (GSCSConstraint *)constraint
+                             forView: (NSView *)view;
+
+- (void) addIntrinsicContentSizeConstraintsToView: (NSView *)view;
+
+- (void) addSupportingInstrictSizeConstraintsToView: (NSView *)view
```
[the remark this answers is not in the document]
Comment (BennyKJohnson) [reply Source/GSAutoLayoutEngine.m:191]: Its mainly to convey that the variable InstrinctWIdth or InstrinctHeight

File: Source/GSAutoLayoutEngine.m [hunk 10]
```diff
@@ -209,14 +650,397 @@ - (void) mapLayoutConstraint: (NSLayoutConstraint *)layoutConstraint
...
+    [GSCSConstraint editConstraintWithVariable: variable];
+  [self addSupportingSolverConstraint: editConstraint
+                  forSolverConstraint: constraint];
+  [self resolveVariableForView: view attribute: attribute];
+}
+
+- (void) addObserverToConstraint: (NSLayoutConstraint *)constranit
+{
+  [constranit addObserver: self
+                forKeyPath:@"constant"
```
[the remark this answers is not in the document]
Comment (BennyKJohnson) [reply Source/GSAutoLayoutEngine.m]: It's actually to do with clang format. Its missing some cases where spaces need to be inserted  :)

File: Source/NSView.m (hunk 3 above)
Comment (BennyKJohnson) [reply Source/NSView.m]: I have moved this method into the private category. The other methods in this category are all public.

File: Source/GSAutoLayoutEngine.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (fredkiefer) [reply Source/GSAutoLayoutEngine.m:191]: I was hinting to the difference between `Instrict` in the method name and `instrinct` in the argument name

File: Headers/AppKit/NSView.h (hunk 2 above)
Comment (fredkiefer) [reply Headers/AppKit/NSView.h:719]: Somehow I cannot see any changes here. Could you please recheck?

File: Source/GSAutoLayoutEngine.m [hunk 11]
```diff
@@ -424,21 +1241,53 @@ - (NSArray *) constraintsForView: (NSView *)view
...
+      return [view baselineOffsetFromBottom];
+    case GSLayoutViewAttributeFirstBaselineOffsetFromTop:
+      return [view firstBaselineOffsetFromTop];
+    case GSLayoutViewAttributeIntrinsicWidth:
+      return [view intrinsicContentSize].width;
+    case GSLayoutViewAttributeInstrinctHeight:
+      return [view intrinsicContentSize].height;
+    default:
+      [[NSException exceptionWithName: @"Not handled"
+                                reason: @"GSLayoutAttribute not handled"
```
Comment (fredkiefer) [inline Source/GSAutoLayoutEngine.m:1270]: The indentation seems wrong here

File: Source/GSAutoLayoutEngine.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (fredkiefer) [reply Source/GSAutoLayoutEngine.m:191]: You fixed the argument name, thank you for that. Did you also look at the method name? Shouldn't that be changed to `Intrinsic` as well?

File: Source/GSAutoLayoutEngine.m [hunk 12]
```diff
@@ -199,8 +346,280 @@ - (void) removeConstraints: (NSArray *)constraints
...
+    [self variableForView: [constraint firstItem]
+             andAttribute: (GSLayoutAttribute)[constraint firstAttribute]];
+  GSCSVariable *secondItemConstraintVariable =
+    [self variableForView: [constraint secondItem]
+             andAttribute: (GSLayoutAttribute)[constraint secondAttribute]];
+
+  CGFloat multiplier = [constraint multiplier];
+
+  GSCSConstraintOperator op = GSCSConstraintOperatorEqual;
+  switch ([constraint relation])
```
Comment (fredkiefer) [inline Source/GSAutoLayoutEngine.m:399]: You could have a map for these values to replace the switch in this and the previous method.

File: Source/GSAutoLayoutEngine.m [hunk 13]
```diff
@@ -199,8 +346,280 @@ - (void) removeConstraints: (NSArray *)constraints
...
+  case GSLayoutViewAttributeFirstBaselineOffsetFromTop:
+    return @"firstBaselineOffsetFromTop";
+  case GSLayoutViewAttributeIntrinsicWidth:
+    return @"intrinsicContentSize.width";
+  case GSLayoutViewAttributeIntrinsicHeight:
+    return @"intrinsicContentSize.height";
+  default:
+    [[NSException
+      exceptionWithName: @"GSLayoutViewAttribute Not handled"
+                  reason: @"The provided GSLayoutViewAttribute does "
```
Comment (fredkiefer) [inline Source/GSAutoLayoutEngine.m:581]: The indentation looks wrong

**Review by fredkiefer — APPROVED** [review]

**Review by fredkiefer — APPROVED** [review]
Perfect

**Pull request merged**